crates/ruff/src/rules/ruff/rules/unnecessary_iterable_allocation_for_first_element.rs

```diff
@@ -175,47 +214,98 @@ fn iterable_name<'a>(expr: &'a Expr, model: &SemanticModel) -> Option<&'a str> {
+                        })
+                    }),
+                Expr::ListComp(ast::ExprListComp {
                     elt, generators, ..
-                })) => generator_iterable(elt, generators),
-                _ => None,
+                }) => match_simple_comprehension(elt, generators)
+                    .map(|range| IterationTarget {
+                        range,
+                        iterable: false,
+                    })
+                    .or_else(|| {
+                        Some(IterationTarget {
+                            range: arg
+                                .range()
+                                // Remove the `[`
+                                .add_start(TextSize::from(1))
+                                // Remove the `]`
+                                .sub_end(TextSize::from(1)),
+                            iterable: true,
+                        })
+                    }),
+                _ => Some(IterationTarget {
+                    range: arg.range(),
+                    iterable: false,
+                }),
             }
         }
         Expr::ListComp(ast::ExprListComp {
             elt, generators, ..
-        }) => generator_iterable(elt, generators),
+        }) => match_simple_comprehension(elt, generators)
+            .map(|range| IterationTarget {
+                range,
+                iterable: false,
+            })
+            .or_else(|| {
+                Some(IterationTarget {
+                    range: expr
+                        .range()
+                        // Remove the `[`
+                        .add_start(TextSize::from(1))
+                        // Remove the `]`
+                        .sub_end(TextSize::from(1)),
+                    iterable: true,
+                })
+            }),
         _ => None,
     }
 }
 
-/// Given a comprehension, returns the name of the iterable over which it iterates, if it's
-/// a simple comprehension (e.g., `x` for `[i for i in x]`).
-fn generator_iterable<'a>(elt: &'a Expr, generators: &'a Vec<Comprehension>) -> Option<&'a str> {
-    // If the `elt` field is anything other than a [`Expr::Name`], we can't be sure that it
-    // doesn't modify the elements of the underlying iterator (e.g., `[i + 1 for i in x][0]`).
-    if !elt.is_name_expr() {
+/// Returns the [`Expr`] target for a comprehension, if the comprehension is "simple"
+/// (e.g., `x` for `[i for i in x]`).
+fn match_simple_comprehension(elt: &Expr, generators: &[Comprehension]) -> Option<TextRange> {
+    let [generator] = generators else {
         return None;
-    }
+    };
 
-    // If there's more than 1 generator, we can't safely say that it fits the diagnostic conditions
-    // (e.g., `[(i, j) for i in x for j in y][0]`).
-    let [generator] = generators.as_slice() else {
+    if generator.is_async {
         return None;
-    };
+    }
 
     // Ignore if there's an `if` statement in the comprehension, since it filters the list.
```

Comment:
Nit: You can go funky on match statements

```suggestion
fn match_simple_comprehension(elt: &Expr, generators: &[Comprehension]) -> Option<TextRange> {
    let [generator @ Comprehension { is_async: false, ... } ] = generators else {
        return None;
    };
```